src/core/ddsi/src/q_entity.c

```diff
@@ -2644,6 +2822,143 @@ static void match_proxy_reader_with_writers (struct proxy_reader *prd, nn_mtime_
   generic_do_match(&prd->e, tnow);
 }
 
+static void match_volatile_secure_endpoints (struct participant *pp, struct proxy_participant *proxypp)
+{
+  struct reader *rd;
+  struct writer *wr;
+  struct proxy_reader *prd;
+  struct proxy_writer *pwr;
+  ddsi_guid_t guid;
+  nn_mtime_t tnow = now_mt ();
+
+  EELOGDISC (&pp->e, "match volatile endpoints (pp "PGUIDFMT") with (proxypp "PGUIDFMT")\n",
+             PGUID(pp->e.guid), PGUID(proxypp->e.guid));
+
+  guid = pp->e.guid;
+  guid.entityid.u = NN_ENTITYID_P2P_BUILTIN_PARTICIPANT_VOLATILE_SECURE_READER;
+  rd = ephash_lookup_reader_guid (pp->e.gv->guid_hash, &guid);
+  assert(rd);
+  guid.entityid.u = NN_ENTITYID_P2P_BUILTIN_PARTICIPANT_VOLATILE_SECURE_WRITER;
+  wr = ephash_lookup_writer_guid (pp->e.gv->guid_hash, &guid);
+  assert(wr);
+
+  guid = proxypp->e.guid;
+  guid.entityid.u = NN_ENTITYID_P2P_BUILTIN_PARTICIPANT_VOLATILE_SECURE_READER;
+  prd = ephash_lookup_proxy_reader_guid (pp->e.gv->guid_hash, &guid);
+  assert(rd);
+  guid.entityid.u = NN_ENTITYID_P2P_BUILTIN_PARTICIPANT_VOLATILE_SECURE_WRITER;
+  pwr = ephash_lookup_proxy_writer_guid (pp->e.gv->guid_hash, &guid);
+  assert(wr);
+
+  if (q_omg_security_match_remote_writer_enabled (rd, pwr))
+  {
+    nn_count_t init_count;
+    reader_add_connection (rd, pwr, &init_count);
+    proxy_writer_add_connection (pwr, rd, tnow, init_count);
+  }
+
+  if (q_omg_security_match_remote_reader_enabled (wr, prd))
+  {
+    proxy_reader_add_connection (prd, wr);
+    writer_add_connection (wr, prd);
+  }
+}
+
+static void match_proxy_writer_with_participant_readers (struct proxy_writer *pwr, struct participant *pp, nn_mtime_t tnow)
```

Comment:
I suspect these are added to be able to match on completion of a handshake. I'd much rather see that done without the code duplication, but obviously iterating over all participant and then scanning all endpoints in this manner scales terribly badly.

There is a solution ... Cyclone would really benefit from more efficient matching in the discovery anyway, and I have some code lying around that would help. I'll try to create a PR for that "real soon now" and then this can be reworked a bit.

Reply:
That change to perform matching more efficiently went into master the other day